src/filters/ES6Ctors.ml

```diff
@@ -0,0 +1,225 @@
+(*
+	The Haxe Compiler
+	Copyright (C) 2005-2019  Haxe Foundation
+
+	This program is free software; you can redistribute it and/or
+	modify it under the terms of the GNU General Public License
+	as published by the Free Software Foundation; either version 2
+	of the License, or (at your option) any later version.
+
+	This program is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with this program; if not, write to the Free Software
+	Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *)
+open Common
+open Globals
+open Type
+open Texpr.Builder
+
+(* name of the method to which the constructor is extracted *)
+let ctor_method_name = "_hx_constructor"
+
+(* name of the static bool flag to skip constructor body execution *)
+let skip_ctor_flag_name = "_hx_skip_constructor"
+
+(* replace super(a,b,c) with super._hx_constructor(a,b,c) *)
+let rec replace_super_call e =
+	match e.eexpr with
+	| TCall ({ eexpr = TConst TSuper } as e_super, args) ->
+		let e_super_hxctor = { e_super with eexpr = TField (e_super, FDynamic ctor_method_name) } in
+		{ e with eexpr = TCall (e_super_hxctor, args) }
+	| _ ->
+		map_expr replace_super_call e
+
+exception Accessed_this of texpr
+
+(* return whether given expression has `this` access before calling `super` *)
+let has_this_before_super e = 
+	let rec loop e =
+		match e.eexpr with 
+		| TCall ({ eexpr = TConst TSuper }, args) ->
+			List.iter loop args;
+			raise Exit
+		| TConst TThis ->
+			raise (Accessed_this e)
+		| _ ->
+			Type.iter loop e
+	in
+	try
+		(loop e; None)
+	with
+		| Exit -> None
+		| Accessed_this e -> Some e
+
+
+(*
+	the filter works in two passes:
+	- mark classes whether they need to support constructor skipping and/or they skip parent's constructor
+	- change the constructors of marked classes (extract body into method and/or add skipping super calls)
+*)
+let rewrite_ctors com =
+	(* we mark classes that need changing by storing them in these two maps *)
+	let needs_ctor_skipping, does_ctor_skipping, inject_super =
+		let l = List.length com.types in
+		Hashtbl.create l, Hashtbl.create l, Hashtbl.create l
+	in
+
+	(*
+		we're using a reference to the root of the inheritance chain so we can easily
+		generate RootClass._hx_skip_constructor expressions
+	*)
+	let rec mark_needs_ctor_skipping cl root_ref p = 
+		(* for non haxe-generated extern classes we can't generate any valid code, so just fail *)
+		if cl.cl_extern && not (Meta.has Meta.HxGen cl.cl_meta) then begin
+			abort "Must call `super()` constructor before accessing `this` in classes derived from an extern class with constructor" p;
+		end;
+		Hashtbl.replace needs_ctor_skipping cl.cl_path root_ref;
```

Comment:
Instead of replacing a value in the hash table and traversing the inheritance chain over and over again for each descendant class found in `com.types` you could retrieve `root_ref` from the hash table and return it immediately if it exists.

Reply:
Right, now I remember what I wanted to look into :)

Reply:
Done!